lib/uksched/include/uk/wait.h

```diff
@@ -39,183 +25,227 @@ extern "C" {
+ */
+#define uk__waitq_lock(wq, flags) \
+	ukplat_spin_lock_irqsave(&((wq)->lock), (flags))
+
+/**
+ * INTERNAL. Release lock on `wq`, restoring irqflags from `flags`.
+ */
+#define uk__waitq_unlock(wq, flags) \
+	ukplat_spin_unlock_irqrestore(&((wq)->lock), (flags))
+
+/**
+ * INTERNAL. Wait using `wq` until `condition` is met or until a deadline.
+ *
+ * Expression returns non-zero if timed out.
+ *
+ * Depending on `condition` may wait 0 or multiple times in the queue.
+ * If `deadline` is non-zero, wait at most until that point in time;
+ * after wakeup compare `deadline` with the value supplied by `clock_expr` to
+ * determine whether a timeout has occurred.
+ * If `lock` evaluates true, it will be released before sleeping with
+ * `unlock_fn` and re-acquired after wakeup with `lock_fn`.
+ */
+#define uk__waitq_wait_until(wq, condition, deadline, clock_expr,	\
+			     lock_fn, unlock_fn, lock)			\
+({									\
+	int uk__waitq_timedout = 0;					\
+									\
+	if (unlikely(!(condition))) {					\
+		struct uk_thread *uk__waitq_current = uk_thread_current();\
+		unsigned long uk__waitq_flags;				\
+									\
+		uk__waitq_lock((wq), uk__waitq_flags);			\
+		if ((condition)) {					\
+			/* Condition changed in the meantime */		\
+			uk__waitq_unlock((wq), uk__waitq_flags);	\
+			goto uk__waitq_wait_out;			\
+		}							\
+		/* Need to wait; add self to waitqueue & block */	\
+		uk__waitq_add((wq), uk__waitq_current);			\
+		for (;;) {						\
+			uk__waitq_block_until(uk__waitq_current, (deadline));\
+			uk__waitq_unlock((wq), uk__waitq_flags);	\
+			if ((lock))					\
+				unlock_fn((lock));			\
+			uk_sched_yield();				\
+			/* Awoken */					\
+			if ((lock))					\
+				lock_fn((lock));			\
+			uk__waitq_lock((wq), uk__waitq_flags);		\
+			if ((condition))				\
+				break;					\
+			if ((deadline) && ((clock_expr) >= (deadline))) {\
+				uk__waitq_timedout = 1;			\
+				break;					\
+			}						\
+			/* Still need to wait; back to sleep */		\
+		}							\
+		/* No more waiting, remove from queue & exit */		\
+		uk__waitq_remove(uk__waitq_current);			\
+		uk__waitq_unlock((wq), uk__waitq_flags);		\
+	}								\
+uk__waitq_wait_out:							\
+	uk__waitq_timedout;						\
 })
 
-#define uk_waitq_wait_deadline(wq, lock_fn, unlock_fn, lock, deadline, \
-			       deadline_condition) \
-({ \
-	struct uk_thread *__current; \
-	unsigned long flags; \
-	int timedout = 0; \
-	DEFINE_WAIT(__wait); \
-	__current = uk_thread_current(); \
-	ukplat_spin_lock_irqsave(&((wq)->sl), flags); \
-	uk_waitq_add(wq, &__wait); \
-	__current->wakeup_time = deadline; \
-	uk_thread_set_blocked(__current); \
-	uk_sched_thread_blocked(__current); \
-	ukplat_spin_unlock_irqrestore(&((wq)->sl), flags); \
-	if (lock) \
-		unlock_fn(lock); \
-	uk_sched_yield(); \
-	if (lock) \
-		lock_fn(lock); \
-	if (deadline_condition) \
-		timedout = 1; \
-	ukplat_spin_lock_irqsave(&((wq)->sl), flags); \
-	/* need to wake up */ \
-	uk_thread_wake(__current); \
```

Comment:
huh, so the `uk_thread_wake` from here and `__wq_wait_event_deadline were redundant?